File: management-service/src/modules/facilities/domain/Address.ts
import { ValueObject } from 'shared/domain';
import { Guard, Result } from 'shared/core';

import { IAddress } from './types';
import { countryCodes } from '../../../resources/countryCodes';

export class Address extends ValueObject<IAddress> {
  public static create(props: IAddress): Result<Address> {
    const nullGuard = Guard.againstNullOrUndefinedBulk([
      {
        argument: props.city,
        argumentName: 'address.city',
      },
      {
        argument: props.countryCode,
        argumentName: 'address.countryCode',
      },
      {
        argument: props.postCode,
        argumentName: 'address.postCode',
      },
      {
        argument: props.street,
        argumentName: 'address.street',
      },
    ]);

    if (!nullGuard.succeeded) {
      return Result.fail(nullGuard);
    }

    const countryCodeGuard = Guard.isOneOf({
      value: props.countryCode,
      argumentName: 'address.countryCode',
      validValues: countryCodes.map((countryCode) => countryCode.code),
    });

    if (!countryCodeGuard.succeeded) {
      return Result.fail(countryCodeGuard);
    }

    return Result.ok(new Address(props));
  }
}

File: community-service/src/Community.Domain/Members/ValueObjects/Address.cs
using System.Collections.Generic;
using Core.Domain;

namespace Community.Domain.Members.ValueObjects
{
    public class Address : ValueObject
    {
        public Address(string city, string postCode, string street)
        {
            City = city;
            PostCode = postCode;
            Street = street;
        }

        public string City { get; }
        public string PostCode { get; }
        public string Street { get; }

        protected override IEnumerable<object> GetEqualityComponents() =>
            new List<object> { City, PostCode, Street };
    }
}
